File: ts/maximam_depth_of_binary_tree.ts
interface TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
}

const createTreeNode = (val?: number, left?: TreeNode | null, right?: TreeNode | null): TreeNode => {
  return {
    val: val === undefined ? 0 : val,
    left: left || null,
    right: right || null,
  };
}

const maxDepth = (root: TreeNode | null): number => {
  if (!root) return 0;

  const left: number = maxDepth(root.left);
  const right: number = maxDepth(root.right);

  return Math.max(left, right) + 1;
}

const rootNode = createTreeNode(3);
const leftNode = createTreeNode(9);
const rightNode = createTreeNode(20);
const rightLeftNode = createTreeNode(15);
const rightRightNode = createTreeNode(7);

rootNode.left = leftNode;
rootNode.right = rightNode;
rightNode.left = rightLeftNode;
rightNode.right = rightRightNode;

const result = maxDepth(rootNode);
console.log(result); // Output: 3

File: easy/ruby/maximam_depth_of_binary_tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def max_depth(root)
  return 0 if root.nil?

  left = max_depth(root.left)
  right = max_depth(root.right)

  [left, right].max + 1
end